src/hugr/ident.rs

```diff
@@ -0,0 +1,98 @@
+use std::borrow::Borrow;
+
+use derive_more::Display;
+use lazy_static::lazy_static;
+use regex::Regex;
+use smol_str::SmolStr;
+use thiserror::Error;
+
+lazy_static! {
+    pub static ref PATH_REGEX: Regex = Regex::new(r"^[\w--\d]\w*(\.[\w--\d]\w*)*$").unwrap();
+}
+
+#[derive(
+    Clone,
+    Debug,
+    Display,
+    PartialEq,
+    Eq,
+    Hash,
+    PartialOrd,
+    Ord,
+    serde::Serialize,
+    serde::Deserialize,
+)]
+/// A non-empty dot-separated list of valid identifiers
+pub struct IdentList(SmolStr);
+
+impl IdentList {
+    /// Makes an IdentList, checking the supplied string is well-formed
+    pub fn new(n: impl Into<SmolStr>) -> Result<Self, InvalidIdentifier> {
+        let n = n.into();
+        if PATH_REGEX.is_match(n.as_str()) {
+            Ok(IdentList(n))
+        } else {
+            Err(InvalidIdentifier(n))
+        }
+    }
+
+    /// Create a new [IdentList] *without* doing the well-formedness check.
+    /// Useful because we want to have static 'const'ants inside the crate,
+    /// but to be used sparingly.
+    pub(crate) const fn new_unchecked(n: &str) -> Self {
+        IdentList(SmolStr::new_inline(n))
+    }
+}
+
+impl Borrow<str> for IdentList {
+    fn borrow(&self) -> &str {
+        self.0.borrow()
+    }
+}
+
+impl std::ops::Deref for IdentList {
+    type Target = str;
+
+    fn deref(&self) -> &str {
+        self.0.deref()
+    }
+}
+
+impl PartialEq<str> for IdentList {
```

Comment:
I think this should be covered by `Deref`, but ok.

Reply:
Sadly not - removing it gives me about 100 errors like this:
```
error[E0277]: can't compare `IdentList` with `str`
   --> src/std_extensions/logic.rs:118:9
    |
118 |         assert_eq!(r.name(), "logic");
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no implementation for `IdentList == str`
```

I started out without this `impl PartialEq` and got everything OK besides the tests, then played around a bit with various Into/Borrow/etc., but couldn't get anything much else to work. Any ideas?

Reply:
This works (without the PartialEq), but is a pain:
```
let n: &str = r.name();
assert_eq!(n, "logic");
```
so that `&str` is critical. IOW, it's exactly the same issue as [the other comment](https://github.com/CQCL-DEV/hugr/pull/491#discussion_r1315606215)

Reply:
>.<
👍

Reply:
Yes, as I think you've already noticed - I've "solved" this (removing the `PartialEq` and the awkward `let`) via a bunch of `as &str`s